docs/INTERNALS.md

```diff
@@ -0,0 +1,336 @@
+custom code generators easier, we don’t guarantee any source or binary compatibility. APIs can
+be added, changed or removed at any time. Therefore, we annotate all publicly available APIs
+with `@ExperimentalAnvilApi`.
+
+## Contributing Bindings / Subcomponents
+
+Earlier it was mentioned that hints are generated in the frontend through code generators and
+the merge phase is implemented in the backend. This is true for contributed modules and component
+interfaces. The exceptions are contributed bindings and subcomponents.
+
+We merge contributed bindings in a `CodeGenerator` in the frontend already, because the generated
+code itself uses Anvil annotations and must be processed by other code generators. The code
+generator for contributed bindings scans the classpath and finds all contributed bindings in
+this module, and then generates one large Dagger module with all binding / provides methods for
+these types. The `CodeGenerator` for contributed subcomponents works very similarly.
+
+## Generate Dagger Factories
+
+This feature is not required to make Anvil work, but it brings significant performance
+improvements at build time. Anvil is a convenience tool and increases build times (more work
+during compilation means longer build times). Generating Dagger factories flipped the coin.
+
+As soon as you use one annotation with importance to Dagger, e.g. `@Inject` or `@Module`, then
+you must enable the Dagger annotation processor and KAPT. KAPT is slow due to all the extra work
+in the stub generating task, annotation processing itself and running javac (Dagger generates
+Java code). The majority of generated code for these annotations is very simple and
+straightforward.
+
+If you opt-into
+[Dagger Factory Generation](https://github.com/square/anvil#dagger-factory-generation), then you
+can safely remove the Dagger annotation processor and likely KAPT as well. Anvil will process
+the annotations Dagger cares about and generate necessary code. Not only does this skip the
+stub generating task and annotation processing, but Anvil also generates Kotlin code and we don’t
+need to run javac additionally (it becomes a no-op). That’s very efficient. Obviously, this
+functionality is implemented through multiple `CodeGenerator`s.
+
+There’s one exception and that’s `@Component`. If Dagger sees `@Component`, then the real magic
+happens and Dagger generates code for the component and all subcomponents. It verifies that all
+dependencies are provided and there are no duplicate bindings. We didn’t replicate that
+functionality within Anvil (then we’d have our own DI framework and wouldn’t need Dagger anymore)
+and these modules must still use KAPT and the Dagger annotation processor. Note that modules
+using `@MergeComponent` must also enable KAPT. However, if you exclusively use subcomponents
+in a module, then the Dagger annotation processor and KAPT aren’t required.
+
+## Debugging Anvil
```

Comment:
This section is super valuable, although it feels like it doesn't quite fit in with the rest in terms of explaining Anvil's internals. I think it's fine to keep here for now but might make sense to eventually split into its own doc focused on tips for an effective workflow for contributors.